### PR TITLE
tiny: fp8 MTP + full-layer XSA → 3.316 val loss

### PR DESCRIPTION
**Pull request by Mister-dev-oss**
# Add fp8 multi-token prediction + full-layer XSA

## Summary

This PR builds on the current Tiny record (learnable [Exclusive Self-Attention (XSA)](https://arxiv.org/pdf/2603.09078) by @CerovazS, #91) and adds two changes to `tiny/train.py`:

1. **fp8 fused-softcapped cross-entropy + multi-token prediction (MTP).** The LM-head loss is computed by an inlined Triton/CUDA kernel that does the softcapped cross-entropy in fp8 (`torch._scaled_mm`) without ever materializing the full `[B·T, 50304]` logits, and predicts 3 future tokens (annealed to zero over training). This is both a denser training signal and a per-step speedup, so the same 3040-step / 16-epoch budget finishes in `13.78m` vs the record's `14.71m`, with ~1.2m of margin to spare.
2. **Extend XSA from `first6` to all layers** (`--xsa-mode all`).

The file is self-contained (the fp8 CE kernel is inlined, no extra `triton_kernels.py`) and every winning hyperparameter is baked as a default, so the standard reproduce command alone reproduces the result:

```bash
python prepare_data.py
torchrun --standalone --nproc_per_node=8 tiny/train.py
```

## What changed

- Inline the fp8 fused-softcapped cross-entropy + MTP kernel directly into `tiny/train.py` (self-contained, no separate kernels module).
- Add `--xsa-mode all` (XSA on every transformer layer) and make it the default.

## Main result

Baseline reference:

| Run | Ckpt avg val loss | Total training time |
| --- | ---: | ---: |
| Current Tiny record (learnable XSA, @CerovazS, #91) | `3.329040` | `14.71m` |

This PR (8×H100, canonical data, default no-flag run, 3040 steps / 16 epochs):

| Run | Setting | Ckpt avg val loss | Total training time |
| --- | --- | ---: | ---: |
| `mtp_fp8_xsa_full` | `torchrun --standalone --nproc_per_node=8 tiny/train.py` (defaults: `--xsa-mode all`) | `3.316223` | `13.78m` |

The run is under the 15-minute training-time budget and beats the current record on checkpoint-averaged validation loss. The result reproduces from the bare command above with no flags.

## Delta

Checkpoint-averaged val loss (`3.316223`):

- vs current record (learnable XSA, @CerovazS, #91): `3.316223 - 3.329040 = -0.012817`

## Validity

- **Comparable metric.** Validation loss is computed with standard `F.cross_entropy`, not the fp8 kernel; fp8 only touches training gradients.
- **Same data and batching.** Canonical FineWeb sample-10BT, both `prepare_data.py` hashes verified (val `6868ed37…`, train `36e7c95c…`); `total_batch_size = 524288` tokens/step, `device_batch_size = 32`, grad-accum 1 — identical to the baseline. `prepare_data.py` is unchanged.
- **Time.** "Total training time" is `13.78m < 15.00m` (wall clock `16.59m`); the README "Time" column is training time, per the record convention.
- **Stable.** No NaNs.

## Epoch validation losses

Run: `mtp_fp8_xsa_full` — default no-flag reproduction (`torchrun --standalone --nproc_per_node=8 tiny/train.py`)

| Epoch | Step | Val loss |
| ---: | ---: | ---: |
| 1 | 190 | `4.842388` |
| 2 | 380 | `4.420628` |
| 3 | 570 | `4.253456` |
| 4 | 760 | `4.162789` |
| 5 | 950 | `4.055982` |
| 6 | 1140 | `3.937681` |
| 7 | 1330 | `3.808964` |
| 8 | 1520 | `3.667390` |
| 9 | 1710 | `3.595037` |
| 10 | 1900 | `3.550258` |
| 11 | 2090 | `3.542891` |
| 12 | 2280 | `3.534469` |
| 13 | 2470 | `3.361121` |
| 14 | 2660 | `3.356654` |
| 15 | 2850 | `3.357803` |
| 16 | 3040 | `3.380340` |

Final metrics:

- Final cutoff validation loss: `3.380340`
- EMA validation loss: `3.333890`
- Checkpoint-averaged validation loss: `3.316223`
- Total training time: `13.78m`
- Wall clock time: `16.59m`
- Peak memory: `63,426 MiB`

## Acknowledgements

This work was also possible thanks to the support provided by Paradigma ([link](https://paradigma.inc/)) and the use of Flywheel ([link](https://flywheel.paradigma.inc/)), Paradigma's infrastructure for research.

Team: @Mister-dev-oss, @CerovazS, @MarioPaerle, @GabrieleCirillo, @crisostomi.

### CODE REVIEW

**Comment by akshayvegesna** [comment]
Thanks! We'll merge this after #92. Could you clean it up so the kernel is in a separate file for clarity?

**Comment by Mister-dev-oss** [comment]
Thanks, done!
